### PR TITLE
v2ray: 4.45.0 -> 5.1.0 and refactor

### PR DESCRIPTION
**Pull request by oxalica**
###### Description of changes

This is a breaking major update: https://github.com/v2fly/v2ray-core/releases/tag/v5.1.0
There are new CLI and upstream systemd units.

The nix file is simplified and refactored, which makes auto-update easier. Upstream now supports user assets overriding under XDG paths since https://github.com/v2fly/v2ray-core/pull/1578. It should be less common to override packaged assets.

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>v2ray</li>
  </ul>
</details>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/v2ray/default.nix [hunk 1]
```diff
@@ -1,69 +1,63 @@
...
-      runHook postBuild
-    '';
+  nativeBuildInputs = [ makeWrapper ];
 
-    installPhase = ''
-      install -Dm755 v2ray v2ctl -t $out/bin
-    '';
+  installPhase = ''
+    runHook preInstall
+    install -Dm555 "$NIX_BUILD_TOP"/go/bin/main $out/bin/v2ray
```
Comment (NickCao) [inline pkgs/tools/networking/v2ray/default.nix:30]: Should avoid using $NIX_BUILD_TOP https://github.com/NixOS/nixpkgs/issues/189691

File: pkgs/tools/networking/v2ray/default.nix [hunk 2]
```diff
@@ -1,69 +1,63 @@
...
   };
 
-in runCommand "v2ray-${version}" {
-  inherit src version;
-  inherit (core) meta;
-
-  nativeBuildInputs = [ makeWrapper ];
+  postFixup = ''
+    wrapProgram $out/bin/v2ray \
+      --set-default V2RAY_LOCATION_ASSET $assetsDrv/share/v2ray
```
Comment (NickCao) [inline pkgs/tools/networking/v2ray/default.nix:43]: v2ray now searches XDG_DATA_DIRS for assets.
`--suffix XDG_DATA_DIRS $assetsDrv/share`

File: pkgs/tools/networking/v2ray/default.nix (hunk 1 above)
Comment (oxalica) [reply pkgs/tools/networking/v2ray/default.nix]: Oops, I just incorrectly read `buildGoPackage`'s code to find this. Fixed now.

File: nixos/modules/services/networking/v2ray.nix [hunk 3]
```diff
@@ -83,13 +83,11 @@ with lib;
...
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/v2ray -config ${configFile}";
-      };
-    };
+    environment.etc."v2ray/config.json".source = configFile;
+
+    systemd.packages = [ cfg.package ];
+
+    # Workaround: https://github.com/NixOS/nixpkgs/issues/81138
+    systemd.services.v2ray.wantedBy = [ "multi-user.target" ];
```
Comment (NickCao) [inline nixos/modules/services/networking/v2ray.nix:91]: Would the service still be restarted/reloaded on configuration changes with this approach? Maybe a reloadTrigger has to be added.

Comment (oxalica) [reply nixos/modules/services/networking/v2ray.nix]: Added `restartTriggers`. It doesn't seem to support online reloading.

**Review by NickCao — APPROVED** [review]
lgtm

**Pull request merged**